Structured data for store/organisation

rahulsingh321 opened the pull request:
Store form UI:
![Screenshot from 2025-02-03 23-31-47](https://github.com/user-attachments/assets/67c7943a-8831-4203-8e22-f8eea2efee11)

fthobe commented on core/db/migrate/20250202173007_add_store_attributes_to_spree_stores.rb:
```diff
@@ -1,11 +1,11 @@
+  def change
+    add_column :spree_stores, :legal_name, :string
+    add_column :spree_stores, :tax_id, :string
+    add_column :spree_stores, :address, :jsonb
```
this field will be hammered if used on home page, I prefer columns, but I wait for @kennyadsl for feedback.
I am unsure about this one: is jsonb much more problematic than regular fields? If someone is smart, the information is placed on an about us page, as google explicitly allows it. Home usually renders better with more success in rich results, but google business can help tremendously there. 

What do you think? jsonb or columns? 

Google on that topic: 

> We recommend placing this information on your home page, or a single page that describes your organization, for example the about us page.

fthobe replied:
Break down address fields

fthobe replied:
I am still not convinced that we should leave jsonb, but I am ok with it if everything else just checks out.

rahulsingh321 replied:
@fthobe I already remove the JSON field and used direct fields instead, resolving this.

fthobe commented:
@rahulsingh321 translations for labels are entirely missing. 
Also here labels are acronyms: 
VAT all capitalized as acronym for Value added tax therefor `VAT ID `
`Tax ID` not, because tax is not an acronym

fthobe commented on backend/app/views/spree/admin/stores/_form.html.erb:
```diff
@@ -81,5 +107,49 @@
+
+      <div class="col-md-6 mb-3">
+        <%= f.label :addressRegion, "Region", class: "form-label" %>
+        <%= text_field_tag 'store[address][addressRegion]', f.object.address&.dig('addressRegion'), class: 'form-control', placeholder: "Region" %>
```
Validate regions as standard by solidus:
```
<%= form_for [:admin, @zone] do |zone_form| %>
  <%= render partial: 'form', locals: { zone_form: zone_form } %>
  <div class="clear"></div>
  <%= render partial: 'spree/admin/shared/edit_resource_links' %>
<% end %>
```

rahulsingh321 replied:
Outdated code reviewed. I took state field instead, not region. Also, zone can be composed of based on multiple countries as well, so using zone here doesn't make sense. Better to use state instead, so used that.

fthobe commented on backend/app/views/spree/admin/stores/_form.html.erb:
```diff
@@ -81,5 +107,49 @@
+      </div>
+
+      <div class="col-md-6 mb-3">
+        <%= f.label :addressRegion, "Region", class: "form-label" %>
```
Translation missing: 
en.yml

 ```
activerecord:
    attributes:
      spree/address:
        address1: Street Address
        address2: Street Address (cont'd)
        city: City
        company: Company
        name: Name
        phone: Phone
        zipcode: Zip Code
```

rahulsingh321 replied:
I used code from the address form itself so its reusing locales automatically. Outdated code reviewed.

fthobe commented on backend/app/views/spree/admin/stores/_form.html.erb:
```diff
@@ -81,5 +107,49 @@
+
+  <div class="col-12">
+    <%= f.label :address, "Address" %>
+    <div class="row border mx-1 py-3 border-secondary rounded">
```
Remove this, has no purpose and design language used nowhere. 

Either use 
`<legend align="center">Store Address</legend>` or remove the box / decorative element totally

rahulsingh321 commented:
Todos:
1: Address optional in store.
2: Need to add the translation
3: Need to add the test cases.

JustShah commented on backend/app/views/spree/admin/stores/_form.html.erb:
```diff
@@ -81,5 +100,84 @@
   </div>
+
+  <fieldset>
+    <legend>Store Address</legend>
```
@rahulsingh321 Just wondering if you can use the existing address partial instead of reintroducing the address columns.

https://github.com/gms-electronics/solidus/blob/main/backend/app/views/spree/admin/shared/_address_form.html.erb

fthobe commented:
@kennyadsl we are unsure if using addresses or just make columns here.

[Specs by google](https://developers.google.com/search/docs/appearance/structured-data/organization) are following: 
> The address (physical or mailing) of your organization, if applicable. Include all properties that apply to your country. The more properties you provide, the higher quality the result is for users. You can provide multiple addresses if you have a location in multiple cities, states, or countries. 

```
<html>
  <head>
    <title>About Us</title>
    <script type="application/ld+json">
    {
      "@context": "https://schema.org",
      "@type": "Organization",
      "url": "https://www.example.com",
      "sameAs": ["https://example.net/profile/example1234", "https://example.org/example1234"],
      "logo": "https://www.example.com/images/logo.png",
      "name": "Example Corporation",
      "description": "The example corporation is well-known for producing high-quality widgets",
      "email": "contact@example.com",
      "telephone": "+47-99-999-9999",
      "address": {
        "@type": "PostalAddress",
        "streetAddress": "Rue Improbable 99",
        "addressLocality": "Paris",
        "addressCountry": "FR",
        "addressRegion": "Ile-de-France",
        "postalCode": "75001"
      },
      "vatID": "FR12345678901",
      "iso6523Code": "0199:724500PMK2A2M1SQQ228"
    }
    </script>
  </head>
  <body>
  </body>
</html>

```

fthobe commented:
@rahulsingh321 ~~shoot at solidusio/solidus (main) as draft and check if tests fail.~~ see below comment

fthobe commented:
@rahulsingh321 wait for PR description by me before making pr

fthobe commented:
@rahulsingh321 please update the screenshot

fthobe commented on api/spec/requests/spree/api/stores_spec.rb:
```diff
@@ -22,6 +26,60 @@ module Spree::Api
+        context "when the country has states" do
+          it "is invalid without a state" do
+            store = Spree::Store.new(name: "Test Store", country: country_with_states, state: nil, url: "spreestore.example.com",
+              mail_from_address: "spreestore.example.com", code: "test-store",)
```
I am not sure this shouldn't be a valid email

fthobe commented on api/spec/requests/spree/api/stores_spec.rb:
```diff
@@ -37,7 +95,20 @@ module Spree::Api
+            "legal_name" => nil,
+            "contact_email" => nil,
+            "description" => nil,
+            "phone" => nil,
```
contact_phone

fthobe commented on api/lib/spree/api_configuration.rb:
```diff
@@ -125,9 +125,10 @@ def promotion_attributes=(value)
-      :bcc_email
+      :id, :name, :legal_name, :url, :meta_description, :meta_keywords, :seo_title,
+      :mail_from_address, :default_currency, :code, :default,
+      :bcc_email, :phone, :contact_email, :tax_id, :vat_id, :description,
```
contact_phone

fthobe commented on backend/app/views/spree/admin/stores/_address_form.html.erb:
```diff
@@ -0,0 +1,65 @@
+    </div>
+
+    <div class="field <%= "#{type}-row" %>">
+      <%= f.label :phone %>
```
Contact Phone

fthobe commented on core/app/models/spree/taxon.rb:
I think @shahmayur001 accidently merged 5618cc651ce85f3c5d5a8ca38cd15b4342e1b652 into this. 
@shahmayur001 please remove it

JustShah replied:
> I think @shahmayur001 accidently merged [5618cc6](https://github.com/gms-electronics/solidus/commit/5618cc651ce85f3c5d5a8ca38cd15b4342e1b652) into this. @shahmayur001 please remove it

@fthobe It's coming from the main branch since my PR got merged. We need to rebase this branch, and it will automatically disappear.

fthobe commented on core/lib/spree/permitted_attributes.rb:
```diff
@@ -113,8 +113,8 @@ module PermittedAttributes
                           :mail_from_address, :cart_tax_country_iso,
-                          :bcc_email, :code, :legal_name, :tax_id, :address,
-                          :contact_phone, :contact_email, :vat_id, :description]
+                          :bcc_email, :code, :legal_name, :tax_id, :contact_phone,
```
Here it's contact phone, we careful to catch all changes.

rahulsingh321 replied:
@fthobe Outdated code reviewed, I will update field name to all over the places from phone to contact_phone.

fthobe reviewed (changes requested):
Generally it looks good, some minor things: 
- try to align the section headers with what they do in the new admin
- try to be more consistent with naming throughout a PR

Rest looks good, form is well structured, tests are there. I don't see why this wouldn't be ok. @shahmayur001 needs to remove the wrongly placed commits from here.

benjaminwil committed:
Add `:order_canceled` event

This event fires when the order `#cancel!` event is fired as part of the
stock order state machine events.

Co-Authored-By: Adam Mueller <adam@super.gd>

benjaminwil committed:
Add `:order_short_shipped` event

This event fires when `OrderCancellations#short_ship` is fired when an
an inventory unit is canceled from an order.

Co-Authored-By: Adam Mueller <adam@super.gd>

benjaminwil committed:
Add `:carton_shipped` event

This adds a `:carton_shipped` event that is published whenever
`OrderShipping#ship` succeeds to mark a carton as shipped.

Co-Authored-By: Adam Mueller <adam@super.gd>

kennyadsl committed:
Merge pull request #6170 from SuperGoodSoft/new-order-events

Add new order events

fthobe committed:
Fix Meta Data Restriction Comment to reflect default setting

The comment was inconsistent with the default setting. 
No functional changes have been made.

tvdeyen committed:
Merge pull request #6171 from fthobe/fix-default-meta-data-config

[Docs] Fix Meta Data Restriction Comment to reflect default setting

swamp09 committed:
[Admin] Fix Unclosed form_tag in table component

This PR fixes an HTML structure break caused by an unclosed form_tag, ensuring that forms added after using ui/table are rendered correctly.

## Background

While customizing the index page to add a new form, I encountered a problem where the added form was not displayed in the browser. Upon investigation, I discovered that the issue stemmed from the form_tag in ui/table/component.html.erb not being explicitly closed, which resulted in subsequent form tags disappearing.

``` rb
<%= render component("ui/table/toolbar").new("data-#{stimulus_id}-target": "batchToolbar", role: "toolbar", "aria-label": t(".batch_actions"), hidden: true) do %>
  <%= form_tag '', id: batch_actions_form_id %>
  <% @data.batch_actions.each do |batch_action| %>
    <%= render_batch_action_button(batch_action) %>
  <% end %>
<% end %>
```

Due to the lack of a proper block structure closing the form_tag, the HTML became malformed, and any form tags declared afterward were not rendered.

Updated the form_tag to use a block form, ensuring that the opening and closing tags are explicitly defined.

rahulsingh321 committed:
[Compliance / SEO] Add structured data fields to Spree stores

This Commit provides the possibility to store legal information in the
store resource. Given the wide array of different information required
by country, the information stored here provides a baseline for most
countries, but might need to be augmented in some jurisdictions. This
commit is not meant as legal advisory. Inform yourself about eventual
requirements in your country of operation.

Some additional information that might be needed:
 - Legal Representative,
 - Share Capital of the company,
 - Contacts for privacy.

Schema.org has a schema to markup organizations that has been widely
adopted by various search engines:

|        | https://schema.org/Organization Support / Docs
|
|--------|-----------------------------------------------------------------------------------|
| Bing   | https://www.bing.com/webmasters/help/?topicid=cc507c09
|
| Google |
https://developers.google.com/search/docs/appearance/structured-data/organization
|
| Yandex |
https://yandex.com/support/webmaster/schema-org/what-is-schema-org.html
|

This update allows to return all organization data on frontend via jsonb
if implemented according to schema documentation on a per store basis
and edit the data via API and Backend (Old / New).

The implemented fields are:

- legal_name,
- contact_email,
- contact_phone,
- vat_id,
- tax_id,
- address1,
- address2,
- Zip code and city,
- state_name,
- country_id,
- and state_id.

Some additional notes regarding the size of the commit:
While this commit seems massive in size, no functionality has been added
apart from storing the values, no logical changes have been made, hence
the commit while bigger in size is simple to analyse. Apart from testing
routines, this commit does not contain functional code. Where possible
the structure of current address forms has been adopted.

Fixes #6173

chaimann committed:
Fix ui/forms/input component for tag: :textarea

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea
In HTML, the initial content of a <textarea> is specified
between its opening and closing tags, not as a `value`
attribute.

Therefore we must set `content`, rather than passing `value` for
 text areas.

tvdeyen committed:
Merge pull request #6174 from chaimann/fix-textarea-component

Fix ui/forms/input component for tag: :textarea

tvdeyen committed:
Merge pull request #6172 from swamp09/fix_unclosed_form_tag_in_admin_table_component

[Admin] Fix Unclosed form_tag in table component

tvdeyen committed:
Revert "GH: Allow backport workflow to trigger actions"

tvdeyen committed:
Merge pull request #6180 from solidusio/revert-6164-backport-actions-permission

Revert "GH: Allow backport workflow to trigger actions"

fthobe commented:
@kennyadsl should be good to go.

fthobe commented:
@kennyadsl could you check again?
